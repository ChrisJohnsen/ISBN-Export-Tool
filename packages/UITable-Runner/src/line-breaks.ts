// Attempt to reproduce line breaks used in a text UITableCell

// This is probably inaccurate, but maybe not entirely useless.

// originally from Scriptable-hosted font measurement
export type LineBreakFontMeasures = {
  enWidth: number,
  spaceWidth: number,
  averageLowercaseWidth: number,
  averageUppercaseWidth: number,
};

const breakAfter = new Set(Array.from(' !-/?|}').map(c => c.codePointAt(0)).flatMap(c => c ? [c] : []));

export type EstimatedLine = { breakCount: number, beforeIndex: number, pointsRemaining: number, lastBreakPoints: number };

export function estimatedLinesForText(fontMeasures: LineBreakFontMeasures): (width: number, text: string, safetyMargin?: number) => EstimatedLine[] {
  return (width, text, safetyMargin = 3 * fontMeasures.enWidth) => {
    type PotentialBreak = { before: number, widthSince: number };
    let potentialBreak: PotentialBreak | null = null;
    const breaks: PotentialBreak[] = [];
    let lineWidth = 0;
    const lines: EstimatedLine[] = [];
    const addLine = (beforeIndex: number, breaks: PotentialBreak[]) => lines.push({
      breakCount: breaks.length,
      beforeIndex,
      pointsRemaining: width - lineWidth,
      lastBreakPoints: (breaks.at(breaks.length - 1)?.widthSince) ?? 0,
    });
    for (let i = 0; i < text.length;) {
      let printing = true;
      let breakpointAfter: boolean | 'forced' = false;
      let chWidth;
      const c = text.codePointAt(i);

      if (c == null) continue;

      if (breakAfter.has(c))
        breakpointAfter = true;

      if (c == 10) {
        // LF
        chWidth = 0;
        printing = false;
        breakpointAfter = 'forced';
      } else if (c == 32) {
        // SP
        chWidth = fontMeasures.spaceWidth;
        printing = false;
      } else if (97 <= c && c <= 122)
        // lowercase
        chWidth = fontMeasures.averageLowercaseWidth;
      else if (65 <= c && c <= 90)
        // uppercase
        chWidth = fontMeasures.averageUppercaseWidth;
      else
        chWidth = fontMeasures.enWidth;

      if (potentialBreak)
        if (lineWidth + potentialBreak.widthSince + chWidth + safetyMargin > width && printing) {
          addLine(potentialBreak.before, breaks.splice(0).slice(0, -1));
          lineWidth = potentialBreak.widthSince + chWidth;
          potentialBreak = null;
        } else
          potentialBreak.widthSince += chWidth;
      else
        if (lineWidth + chWidth + safetyMargin > width && printing) {
          addLine(i, []);
          lineWidth = chWidth;
        } else
          lineWidth += chWidth;

      i += String.fromCodePoint(c).length;
      if (breakpointAfter == 'forced') {
        addLine(i, breaks.splice(0));
        lineWidth = 0;
        potentialBreak = null;
      } else if (breakpointAfter) {
        if (potentialBreak)
          lineWidth += potentialBreak.widthSince;
        potentialBreak = { before: i, widthSince: 0 };
        breaks.push(potentialBreak);
      }
    }
    if (potentialBreak)
      lineWidth += potentialBreak.widthSince;
    addLine(text.length, breaks);
    return lines;
  };
}
