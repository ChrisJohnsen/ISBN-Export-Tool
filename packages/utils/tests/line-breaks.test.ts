import { describe, test, expect } from '@jest/globals';
import { estimatedLinesForText } from 'utils/line-breaks.js';

const fakeFm = {
  enWidth: 1,
  spaceWidth: 0.5,
  averageLowercaseWidth: 0.75,
  averageUppercaseWidth: 1.25,
};

describe('estimatedLinesForText', () => {
  test('empty string is one line', () => {
    const lines = estimatedLinesForText(fakeFm)(10, '');

    expect(lines).toHaveLength(1);
    expect(lines[0].beforeIndex).toEqual(0);
  });

  test('single line', () => {
    const str = 'This is some text.';
    const lines = estimatedLinesForText(fakeFm)(17, str);

    expect(lines).toHaveLength(1);
    expect(lines).toMatchSnapshot(str);
  });

  test('normal break', () => {
    const str = 'This is longer text. It will break onto two lines.';
    const lines = estimatedLinesForText(fakeFm)(23, str);

    expect(lines).toHaveLength(2);
    expect(lines).toMatchSnapshot(str);
  });

  test('two lines, no extra breaks', () => {
    const str = 'This has two lines.\nThis is the second.';
    const lines = estimatedLinesForText(fakeFm)(20, str);

    expect(lines).toHaveLength(2);
    expect(lines).toMatchSnapshot(str);
  });

  test('two lines, both breaking', () => {
    const str = 'This is the first long line. It will break.\nThe second line will also break into two.';
    const lines = estimatedLinesForText(fakeFm)(20, str);

    expect(lines).toHaveLength(4);
    expect(lines).toMatchSnapshot(str);
  });

  test('underscore is not a preferred break point', () => {
    const str = 'An_underscore?';
    const lines = estimatedLinesForText(fakeFm)(10, str);

    expect(lines).toHaveLength(2);
    expect(lines[0].breakCount).toEqual(0);
    expect(lines[0].pointsRemaining).toBeLessThanOrEqual(3 + fakeFm.averageLowercaseWidth); // break should be in "underscore": lots of lowercase letters
    expect(lines).toMatchSnapshot(str);
  });

  test.each(Array.from(' !-/?|}'))('breaking character: %s', c => {
    const str = '_____' + c + '_____';
    const lines = estimatedLinesForText(fakeFm)(10, str);

    expect(lines).toHaveLength(2);
    expect(lines[0].beforeIndex).toEqual(6);
    expect(lines).toMatchSnapshot(str);
  });

  test('pseudo-realistic', () => {
    const lines = estimatedLinesForText({
      // Large (100%), bold; measured on a 2x scaled device
      enWidth: 9.5,
      spaceWidth: 4.1,
      averageLowercaseWidth: 8.711538461538462,
      averageUppercaseWidth: 11.423076923076923,
    })(374, 'This is a semi-realistic example that uses a realistic font measure set.\nIt has a "hard" line break. It also has lots of text that wil flow over multiple display lines.');

    expect(lines).toHaveLength(5);
    expect(lines
      .map(line =>
      ({
        ...line,
        lastBreakPoints: line.lastBreakPoints.toFixed(3), // these are not integers, so round them to a few digits and check them as strings
        pointsRemaining: line.pointsRemaining.toFixed(3),
      })))
      .toMatchSnapshot();
  });
});
