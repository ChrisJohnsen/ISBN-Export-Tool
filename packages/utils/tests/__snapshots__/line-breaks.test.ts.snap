// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`estimatedLinesForText breaking character:  : _____ _____ 1`] = `
[
  {
    "beforeIndex": 6,
    "breakCount": 0,
    "lastBreakPoints": 0,
    "pointsRemaining": 4.5,
  },
  {
    "beforeIndex": 11,
    "breakCount": 0,
    "lastBreakPoints": 0,
    "pointsRemaining": 5,
  },
]
`;

exports[`estimatedLinesForText breaking character: !: _____!_____ 1`] = `
[
  {
    "beforeIndex": 6,
    "breakCount": 0,
    "lastBreakPoints": 0,
    "pointsRemaining": 4,
  },
  {
    "beforeIndex": 11,
    "breakCount": 0,
    "lastBreakPoints": 0,
    "pointsRemaining": 5,
  },
]
`;

exports[`estimatedLinesForText breaking character: /: _____/_____ 1`] = `
[
  {
    "beforeIndex": 6,
    "breakCount": 0,
    "lastBreakPoints": 0,
    "pointsRemaining": 4,
  },
  {
    "beforeIndex": 11,
    "breakCount": 0,
    "lastBreakPoints": 0,
    "pointsRemaining": 5,
  },
]
`;

exports[`estimatedLinesForText breaking character: ?: _____?_____ 1`] = `
[
  {
    "beforeIndex": 6,
    "breakCount": 0,
    "lastBreakPoints": 0,
    "pointsRemaining": 4,
  },
  {
    "beforeIndex": 11,
    "breakCount": 0,
    "lastBreakPoints": 0,
    "pointsRemaining": 5,
  },
]
`;

exports[`estimatedLinesForText breaking character: |: _____|_____ 1`] = `
[
  {
    "beforeIndex": 6,
    "breakCount": 0,
    "lastBreakPoints": 0,
    "pointsRemaining": 4,
  },
  {
    "beforeIndex": 11,
    "breakCount": 0,
    "lastBreakPoints": 0,
    "pointsRemaining": 5,
  },
]
`;

exports[`estimatedLinesForText breaking character: }: _____}_____ 1`] = `
[
  {
    "beforeIndex": 6,
    "breakCount": 0,
    "lastBreakPoints": 0,
    "pointsRemaining": 4,
  },
  {
    "beforeIndex": 11,
    "breakCount": 0,
    "lastBreakPoints": 0,
    "pointsRemaining": 5,
  },
]
`;

exports[`estimatedLinesForText breaking character: -: _____-_____ 1`] = `
[
  {
    "beforeIndex": 6,
    "breakCount": 0,
    "lastBreakPoints": 0,
    "pointsRemaining": 4,
  },
  {
    "beforeIndex": 11,
    "breakCount": 0,
    "lastBreakPoints": 0,
    "pointsRemaining": 5,
  },
]
`;

exports[`estimatedLinesForText normal break: This is longer text. It will break onto two lines. 1`] = `
[
  {
    "beforeIndex": 24,
    "breakCount": 4,
    "lastBreakPoints": 2.5,
    "pointsRemaining": 5,
  },
  {
    "beforeIndex": 50,
    "breakCount": 4,
    "lastBreakPoints": 4.75,
    "pointsRemaining": 4.25,
  },
]
`;

exports[`estimatedLinesForText pseudo-realistic 1`] = `
[
  {
    "beforeIndex": 43,
    "breakCount": 7,
    "lastBreakPoints": "38.946",
    "pointsRemaining": "28.185",
  },
  {
    "beforeIndex": 73,
    "breakCount": 4,
    "lastBreakPoints": "35.635",
    "pointsRemaining": "174.658",
  },
  {
    "beforeIndex": 113,
    "breakCount": 8,
    "lastBreakPoints": "30.235",
    "pointsRemaining": "59.254",
  },
  {
    "beforeIndex": 154,
    "breakCount": 7,
    "lastBreakPoints": "73.792",
    "pointsRemaining": "53.719",
  },
  {
    "beforeIndex": 168,
    "breakCount": 1,
    "lastBreakPoints": "53.058",
    "pointsRemaining": "255.862",
  },
]
`;

exports[`estimatedLinesForText single line: This is some text. 1`] = `
[
  {
    "beforeIndex": 18,
    "breakCount": 3,
    "lastBreakPoints": 4,
    "pointsRemaining": 3.5,
  },
]
`;

exports[`estimatedLinesForText two lines, both breaking: This is the first long line. It will break.
The second line will also break into two. 1`] = `
[
  {
    "beforeIndex": 23,
    "breakCount": 4,
    "lastBreakPoints": 3.5,
    "pointsRemaining": 3.5,
  },
  {
    "beforeIndex": 44,
    "breakCount": 3,
    "lastBreakPoints": 4.75,
    "pointsRemaining": 9.5,
  },
  {
    "beforeIndex": 65,
    "breakCount": 3,
    "lastBreakPoints": 3.5,
    "pointsRemaining": 4.75,
  },
  {
    "beforeIndex": 85,
    "breakCount": 3,
    "lastBreakPoints": 3.25,
    "pointsRemaining": 5.5,
  },
]
`;

exports[`estimatedLinesForText two lines, no extra breaks: This has two lines.
This is the second. 1`] = `
[
  {
    "beforeIndex": 20,
    "breakCount": 3,
    "lastBreakPoints": 4.75,
    "pointsRemaining": 10.5,
  },
  {
    "beforeIndex": 39,
    "breakCount": 3,
    "lastBreakPoints": 5.5,
    "pointsRemaining": 5.75,
  },
]
`;

exports[`estimatedLinesForText underscore is not a preferred break point: An_underscore? 1`] = `
[
  {
    "beforeIndex": 8,
    "breakCount": 0,
    "lastBreakPoints": 0,
    "pointsRemaining": 3.25,
  },
  {
    "beforeIndex": 14,
    "breakCount": 1,
    "lastBreakPoints": 0,
    "pointsRemaining": 5.25,
  },
]
`;
